### ISSUE TITLE
Some Questions about the Target Label

### ISSUE THREAD

**Issue by lylaiyy** [xuxiong0214/BTIDBF#3]
![Image](https://github.com/user-attachments/assets/87869066-f8d2-41bd-9cc9-01cb294b8b32)
As shown in the above figure, I used these default parameters when reproducing.

I injected 5 * 5 color triggers on the CIFAR10 dataset, set the target label to 4, and saved the poisoned ResNet18 model. Then I tested it using the code in btidbf. py.

The operation result of feature decoupling is shown in the following figure：
![Image](https://github.com/user-attachments/assets/6429c14d-5f05-4192-b593-f6e9d9da117b)

The running result of the backdoor sample generator is shown in the following figure. I modified the code to evaluate the target label every time I optimized it, and the result showed that the target label was 9.

![Image](https://github.com/user-attachments/assets/0b06fb1f-e74a-43ee-81a2-402820ac039a)

![Image](https://github.com/user-attachments/assets/04b03e2d-0358-4fe2-9fef-acf9c754512b)

I don't know where the problem lies. My code is as follows:

`import os, sys

# 取得当前脚本所在目录的父目录
parent_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
# 插到模块搜索路径最前面
sys.path.insert(0, parent_dir)
print(parent_dir)

from Nets import CNNFMNIST
from utils.options import args_parser
import shap
import numpy as np
import torch
from torchvision import datasets, transforms
import matplotlib.pyplot as plt
from torch.utils.data import DataLoader
from unet_model import UNet
from tqdm import tqdm
from mask import MaskGenerator
from get_tar_label import get_target_label
import random
import copy
from Test import test_ba
from resnet import ResNet18
from models.helper import get_denormalizer, get_normalizer

torch.manual_seed(43)
random.seed(43)
np.random.seed(43)
if torch.cuda.is_available():
        torch.cuda.manual_seed(43)
        torch.cuda.manual_seed_all(43)

if __name__ == '__main__':
    args = args_parser()
    args.device = torch.device('cuda:{}'.format(args.gpu) if torch.cuda.is_available() and args.gpu != -1 else 'cpu')
    if args.dataset == 'fmnist':
        transform_train = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])
        transform_test = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])
        dataset_train = datasets.FashionMNIST('./data/fmnist', train=True, download=True, transform=transform_train)
        dataset_test = datasets.FashionMNIST('./data/fmnist', train=False, download=True, transform=transform_test)

        all_indices = list(range(len(dataset_test)))
        random_indices = random.sample(all_indices, 1000)
        ba_dataset = torch.utils.data.DataLoader(dataset_test,
                                                 batch_size=64,
                                                 sampler=torch.utils.data.sampler.SubsetRandomSampler(random_indices)
                                                 )
        model_path = './model_save/fmnist_0.2_global_model_epoch_250.pth'
        classifier = CNNFMNIST(args)
        classifier.load_state_dict(torch.load(model_path, map_location=torch.device('cpu')))
        _, asr = test_ba(classifier, ba_dataset, args=args)
        print(f"ASR: {asr}")
        device = args.device
        classifier.to(device)
        bd_gen = UNet(n_channels=1, num_classes=1, base_filter_num=32, num_blocks=2)
        # bd_gen.load_state_dict(
        #     torch.load(os.path.join("./fmnist_init_generator.pt"), map_location=torch.device('cpu')))

    elif args.dataset == 'cifar':
        transform_train = transforms.Compose([
            transforms.RandomCrop(32, padding=4),
            transforms.RandomHorizontalFlip(0.5),
            transforms.ToTensor(),
            transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
        ])

        transform_test = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
        ])

        dataset_train = datasets.CIFAR10('./data/cifar', train=True, download=True, transform=transform_train)
        dataset_test = datasets.CIFAR10('./data/cifar', train=False, download=True, transform=transform_test)

        all_indices = list(range(len(dataset_test)))
        random_indices = random.sample(all_indices, 1000)
        ba_dataset = torch.utils.data.DataLoader(dataset_test,
                                                 batch_size=64,
                                                 sampler=torch.utils.data.sampler.SubsetRandomSampler(random_indices)
                                                 )
        model_path = './model_save/cifar_0.2_global_model_epoch_250.pth'
        classifier = ResNet18()
        classifier.load_state_dict(torch.load(model_path, map_location=torch.device('cpu')))
        asr = test_ba(classifier, ba_dataset, args=args)
        print(f"ASR: {asr}")
        device = args.device
        classifier.to(device)
        bd_gen = UNet(n_channels=3, num_classes=3, base_filter_num=32, num_blocks=4)
        bd_gen.load_state_dict(
            torch.load(os.path.join("./cifar_init_generator.pt"), map_location=torch.device('cpu')))

    # all_indices_test = list(range(len(dataset_test)))
    # random_indices = random.sample(all_indices_test, 300)

    detect_label_dataset = torch.utils.data.DataLoader(dataset_test, batch_size=args.bs)

    all_indices_train = list(range(len(dataset_train)))
    random_indices = random.sample(all_indices_train, 600)
    cln_trainloader = torch.utils.data.DataLoader(dataset_train,
                                                  batch_size=args.bs,
                                                  sampler=torch.utils.data.sampler.SubsetRandomSampler(random_indices)
                                                  )

    bd_gen = bd_gen.to(device)
    opt_bd = torch.optim.Adam(bd_gen.parameters(), lr=args.gen_lr)
    bd_gen.eval()

    mse = torch.nn.MSELoss()
    ce = torch.nn.CrossEntropyLoss()
    if args.dataset == 'fmnist':
        tmp_img = torch.ones([1, 1, 28, 28], device=device)
    else:
        tmp_img = torch.ones([1, 3, 32, 32], device=device)
    tmp_feat = classifier.from_input_to_features(tmp_img).detach()
    feat_shape = tmp_feat.shape
    init_mask = torch.randn(feat_shape).to(device)
    m_gen = MaskGenerator(init_mask=init_mask, classifier=classifier)
    opt_m = torch.optim.Adam([m_gen.mask_tanh], lr=0.01)
    max_m_feat = float('inf')
    final_m_gen = None
    for m in range(args.mround):
        tloss = 0
        tloss_pos_pred = 0
        tloss_neg_pred = 0
        m_gen.train()
        classifier.eval()
        pbar = tqdm(cln_trainloader, desc="Decoupling Benign Features")
        for batch_idx, (cln_img, targets) in enumerate(pbar):
            opt_m.zero_grad()
            cln_img = cln_img.to(device)
            targets = targets.to(device)
            feat_mask = m_gen.get_raw_mask()

            # with torch.no_grad():
            #     pred_label = classifier(cln_img).argmax(1)
            # good_mask = (pred_label == targets)
            # if good_mask.sum() == 0:
            #     continue
            cln_feat = classifier.from_input_to_features(cln_img)
            mask_pos_pred = classifier.from_features_to_output(feat_mask * cln_feat)
            demask_neg_pred = classifier.from_features_to_output((1 - feat_mask) * cln_feat)
            mask_norm = torch.norm(feat_mask, 1)

            loss_pos_pred = ce(mask_pos_pred, targets)
            loss_neg_pred = ce(demask_neg_pred, targets)
            loss = loss_pos_pred - loss_neg_pred

            loss.backward()
            opt_m.step()

            tloss += loss.item()
            tloss_pos_pred += loss_pos_pred.item()
            tloss_neg_pred += loss_neg_pred.item()
            pbar.set_postfix({"epoch": "{:d}".format(m),
                              "loss": "{:.4f}".format(tloss / (batch_idx + 1)),
                              "loss_pos_pred": "{:.4f}".format(tloss_pos_pred / (batch_idx + 1)),
                              "loss_neg_pred": "{:.4f}".format(tloss_neg_pred / (batch_idx + 1)),
                              "mask_norm": "{:.4f}".format(mask_norm)})
    #     if tloss_pos_pred < max_m_feat:
    #         max_m_feat = tloss_pos_pred
    #         final_m_gen = copy.deepcopy(m_gen)
    #         print('-1')
    #
    # feat_mask = final_m_gen.get_raw_mask().detach()
    feat_mask = m_gen.get_raw_mask().detach()
    max_bd_feat = -float('inf')
    final_bd_gen = None

    for u in range(args.uround):
        tloss = 0
        tloss_benign_feat = 0
        tloss_backdoor_feat = 0
        tloss_norm = 0

        m_gen.eval()
        bd_gen.train()
        classifier.eval()
        pbar = tqdm(cln_trainloader, desc="Training Backdoor Generator")
        for batch_idx, (cln_img, targets) in enumerate(pbar):
            cln_img = cln_img.to(device)
            bd_gen_img = bd_gen(cln_img)
            cln_feat = classifier.from_input_to_features(cln_img)
            bd_gen_feat = classifier.from_input_to_features(bd_gen_img)
            loss_benign_feat = mse(feat_mask * cln_feat, feat_mask * bd_gen_feat)
            loss_backdoor_feat = mse((1 - feat_mask) * cln_feat, (1 - feat_mask) * bd_gen_feat)
            loss_norm = mse(cln_img, bd_gen_img)

            if loss_norm > args.norm_bound or loss_benign_feat > args.feat_bound:
                loss = loss_norm
            else:
                loss = -loss_backdoor_feat + 0.01 * loss_benign_feat

            # perceptual_weight = max(0.01, 0.1 - u*0.005)  # 随训练递减
            # loss = (1.0 * loss_benign_feat +
            #         -2.0 * loss_backdoor_feat +
            #         0.5 * loss_norm +
            #         perceptual_weight * loss_norm)
            opt_bd.zero_grad()
            loss.backward()
            opt_bd.step()

            tloss += loss.item()
            tloss_benign_feat += loss_benign_feat.item()
            tloss_backdoor_feat += loss_backdoor_feat.item()
            tloss_norm += loss_norm.item()
            pbar.set_postfix({"epoch": "{:d}".format(u),
                              "loss": "{:.4f}".format(tloss / (batch_idx + 1)),
                              "loss_bengin_feat": "{:.4f}".format(tloss_benign_feat / (batch_idx + 1)),
                              "loss_backdoor_feat": "{:.4f}".format(tloss_backdoor_feat / (batch_idx + 1)),
                              "loss_norm": "{:.4f}".format(tloss_norm / (batch_idx + 1))})

    #     if args.use_max_bd_feat and tloss_backdoor_feat > max_bd_feat:
    #         max_bd_feat = tloss_backdoor_feat
    #         final_bd_gen = copy.deepcopy(bd_gen)
    #
    # if args.use_max_bd_feat and not max_bd_feat is None:
    #     bd_gen = final_bd_gen

        detected_tlabel = get_target_label(testloader=cln_trainloader, testmodel=classifier, midmodel=copy.deepcopy(bd_gen))
        print(f"Target Label is {detected_tlabel}")`

**Comment by lylaiyy** [comment]
I have identified the problem. I found that when decoupling benign features, my code is:
` m_gen.train()`
 `classifier.eval()`
The original code is:
` m_gen.train()`
 `classifier.train()`

However, when benign features are decoupled, shouldn't the model be frozen and then masked? Why use train() instead of eval()? And using eval() will result in reverse failure. I would greatly appreciate it if you could explain this issue.
